File: windows/src/App/Net/TransferService.cs
using System.IO;
using System.Net;
using System.Net.Sockets;

namespace CompartirArchivosRED.Net;

public sealed class IncomingInfo
{
    public string SenderName = "";
    public int FileCount;
    public long TotalSize;
    public string Pin = "";
}

/// <summary>Servidor de recepción (TCP 45455) + cliente de envío. Ver PROTOCOL.md.</summary>
public sealed class TransferService : IDisposable
{
    private readonly string _selfName;
    private TcpListener? _listener;
    private CancellationTokenSource? _cts;

    public string DownloadFolder { get; private set; }

    public event Action<string>? Log;
    public event Action<double>? Progress;              // 0..1
    public event Action<IncomingInfo>? IncomingStarted; // mostrar PIN al usuario
    public event Action<bool>? IncomingFinished;        // éxito?

    private static string DefaultFolder => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
        "Downloads", "CompartirArchivosRED");

    private static string SettingsFile => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "CompartirArchivosRED", "download-folder.txt");

    public TransferService(string selfName)
    {
        _selfName = selfName;
        DownloadFolder = LoadFolder();
        try { Directory.CreateDirectory(DownloadFolder); }
        catch { DownloadFolder = DefaultFolder; Directory.CreateDirectory(DownloadFolder); }
    }

    private static string LoadFolder()
    {
        try
        {
            if (File.Exists(SettingsFile))
            {
                var saved = File.ReadAllText(SettingsFile).Trim();
                if (!string.IsNullOrWhiteSpace(saved)) return saved;
            }
        }
        catch { }
        return DefaultFolder;
    }

    /// <summary>Cambia la carpeta de descargas y la persiste. Devuelve true si OK.</summary>
    public bool ChangeDownloadFolder(string path)
    {
        try
        {
            Directory.CreateDirectory(path);
            DownloadFolder = path;
            Directory.CreateDirectory(Path.GetDirectoryName(SettingsFile)!);
            File.WriteAllText(SettingsFile, path);
            Log?.Invoke($"Carpeta de descargas: {path}");
            return true;
        }
        catch (Exception ex)
        {
            Log?.Invoke("No se pudo cambiar la carpeta: " + ex.Message);
            return false;
        }
    }

    public void StartServer()
    {
        _cts = new CancellationTokenSource();
        _listener = new TcpListener(IPAddress.Any, Proto.TransferPort);
        _listener.Start();
        Task.Run(() => AcceptLoop(_cts.Token));
    }

    private async Task AcceptLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                var client = await _listener!.AcceptTcpClientAsync(ct);
                _ = Task.Run(() => HandleIncoming(client));
            }
            catch (OperationCanceledException) { break; }
            catch { }
        }
    }

    private void HandleIncoming(TcpClient client)
    {
        bool ok = false;
        bool announced = false;
        try
        {
            using (client)
            using (var stream = client.GetStream())
            {
                var f = new Framed(stream);
                var offer = Msg.FromJson(f.ReadLine() ?? "");
                if (offer == null || offer.Type != Proto.OFFER) return;

                string pin = Random.Shared.Next(0, 1_000_000).ToString("D6");
                var info = new IncomingInfo
                {
                    SenderName = offer.Name ?? "?",
                    FileCount = offer.FileCount ?? 0,
                    TotalSize = offer.TotalSize ?? 0,
                    Pin = pin
                };
                announced = true;
                IncomingStarted?.Invoke(info);
                Log?.Invoke($"Solicitud de «{info.SenderName}» ({info.FileCount} archivo/s). PIN: {pin}");

                var expires = DateTime.UtcNow.AddSeconds(60);
                f.WriteLine(new Msg { Type = Proto.PIN_REQUIRED }.ToJson());

                var pinMsg = Msg.FromJson(f.ReadLine() ?? "");
                if (pinMsg == null || pinMsg.Type != Proto.PIN)
                { f.WriteLine(new Msg { Type = Proto.REJECTED, Reason = "denied" }.ToJson()); return; }
                if (DateTime.UtcNow > expires)
                { f.WriteLine(new Msg { Type = Proto.REJECTED, Reason = "expired" }.ToJson()); Log?.Invoke("PIN expirado."); return; }
                if (pinMsg.Pin != pin)
                { f.WriteLine(new Msg { Type = Proto.REJECTED, Reason = "bad_pin" }.ToJson()); Log?.Invoke("PIN incorrecto."); return; }

                f.WriteLine(new Msg { Type = Proto.ACCEPTED }.ToJson());

                long total = info.TotalSize > 0 ? info.TotalSize : 1;
                long done = 0;
                while (true)
                {
                    var line = f.ReadLine();
                    if (line == null) break;
                    var m = Msg.FromJson(line);
                    if (m == null) break;
                    if (m.Type == Proto.DONE) { ok = true; break; }
                    if (m.Type != Proto.FILE) continue;

                    string name = SanitizeName(m.Name ?? "archivo");
                    long size = m.Size ?? 0;
                    string dest = UniquePath(Path.Combine(DownloadFolder, name));
                    long baseDone = done;
                    using (var fs = File.Create(dest))
                    {
                        f.CopyExact(fs, size, written =>
                            Progress?.Invoke(Math.Min(1.0, (baseDone + written) / (double)total)));
                    }
                    done += size;
                    Log?.Invoke($"Recibido: {name} ({FormatSize(size)})");
                }
            }
        }
        catch (Exception ex) { Log?.Invoke("Error en recepción: " + ex.Message); }
        finally
        {
            Progress?.Invoke(0);
            if (announced) IncomingFinished?.Invoke(ok);
        }
    }

    public async Task<bool> SendFilesAsync(Peer peer, IEnumerable<string> paths, Func<Task<string?>> pinProvider)
    {
        var files = paths.Where(File.Exists).ToList();
        if (files.Count == 0) { Log?.Invoke("No hay archivos válidos que enviar."); return false; }
        long totalSize = files.Sum(p => new FileInfo(p).Length);

        try
        {
            using var client = new TcpClient();
            await client.ConnectAsync(peer.Address, peer.Port);
            using var stream = client.GetStream();
            var f = new Framed(stream);

            f.WriteLine(new Msg
            {
                Type = Proto.OFFER,
                Name = _selfName,
                FileCount = files.Count,
                TotalSize = totalSize
            }.ToJson());

            var req = Msg.FromJson(f.ReadLine() ?? "");
            if (req == null || req.Type != Proto.PIN_REQUIRED)
            { Log?.Invoke("El receptor no solicitó PIN."); return false; }

            string? pin = await pinProvider();
            if (string.IsNullOrWhiteSpace(pin)) { Log?.Invoke("Envío cancelado."); return false; }
            f.WriteLine(new Msg { Type = Proto.PIN, Pin = pin!.Trim() }.ToJson());

            var resp = Msg.FromJson(f.ReadLine() ?? "");
            if (resp == null || resp.Type != Proto.ACCEPTED)
            {
                Log?.Invoke("Rechazado por el receptor: " + (resp?.Reason ?? "desconocido"));
                return false;
            }

            long done = 0;
            long total = totalSize > 0 ? totalSize : 1;
            foreach (var path in files)
            {
                var fi = new FileInfo(path);
                f.WriteLine(new Msg { Type = Proto.FILE, Name = fi.Name, Size = fi.Length }.ToJson());
                long baseDone = done;
                using (var fs = File.OpenRead(path))
                {
                    f.WriteBytes(fs, fi.Length, sent =>
                        Progress?.Invoke(Math.Min(1.0, (baseDone + sent) / (double)total)));
                }
                done += fi.Length;
                Log?.Invoke($"Enviado: {fi.Name} ({FormatSize(fi.Length)})");
            }
            f.WriteLine(new Msg { Type = Proto.DONE }.ToJson());
            Log?.Invoke("Transferencia completada.");
            return true;
        }
        catch (Exception ex) { Log?.Invoke("Error en envío: " + ex.Message); return false; }
        finally { Progress?.Invoke(0); }
    }

    private static string SanitizeName(string name)
    {
        name = Path.GetFileName(name);
        foreach (var c in Path.GetInvalidFileNameChars()) name = name.Replace(c, '_');
        return string.IsNullOrWhiteSpace(name) ? "archivo" : name;
    }

    private static string UniquePath(string path)
    {
        if (!File.Exists(path)) return path;
        string dir = Path.GetDirectoryName(path)!;
        string name = Path.GetFileNameWithoutExtension(path);
        string ext = Path.GetExtension(path);
        int i = 1;
        string candidate;
        do { candidate = Path.Combine(dir, $"{name} ({i}){ext}"); i++; }
        while (File.Exists(candidate));
        return candidate;
    }

    public static string FormatSize(long bytes)
    {
        string[] u = { "B", "KB", "MB", "GB", "TB" };
        double v = bytes; int i = 0;
        while (v >= 1024 && i < u.Length - 1) { v /= 1024; i++; }
        return $"{v:0.#} {u[i]}";
    }

    public void Dispose()
    {
        try { _cts?.Cancel(); } catch { }
        try { _listener?.Stop(); } catch { }
    }
}

File: android/app/src/main/java/com/compartirarchivosred/app/net/TransferService.kt
package com.compartirarchivosred.app.net

import android.content.Context
import android.net.Uri
import android.os.Environment
import android.provider.OpenableColumns
import android.webkit.MimeTypeMap
import androidx.documentfile.provider.DocumentFile
import com.compartirarchivosred.app.model.IncomingInfo
import com.compartirarchivosred.app.model.Peer
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.util.Random
import kotlin.concurrent.thread

/** Servidor de recepción (TCP 45455) + cliente de envío. Ver PROTOCOL.md. */
class TransferService(
    private val context: Context,
    private val selfName: String,
    private val onLog: (String) -> Unit,
    private val onProgress: (Float) -> Unit,
    private val onIncoming: (IncomingInfo?) -> Unit
) {
    @Volatile private var running = false
    private var serverSocket: ServerSocket? = null

    /** Carpeta de recepción elegida por el usuario (SAF). Si es null, se usa [downloadDir]. */
    @Volatile var receiveTreeUri: Uri? = null

    val downloadDir: File =
        (context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)
            ?: File(context.filesDir, "Downloads")).apply { mkdirs() }

    fun startServer() {
        if (running) return
        running = true
        thread(isDaemon = true) {
            try {
                val ss = ServerSocket(Proto.TRANSFER_PORT)
                serverSocket = ss
                while (running) {
                    val client = ss.accept()
                    thread(isDaemon = true) { handleIncoming(client) }
                }
            } catch (e: Exception) {
                if (running) onLog("Servidor detenido: ${e.message}")
            }
        }
    }

    private fun handleIncoming(client: Socket) {
        var ok = false
        var announced = false
        try {
            val f = Framed(client.getInputStream(), client.getOutputStream())
            val offerLine = f.readLine() ?: return
            val offer = JSONObject(offerLine)
            if (offer.optString("type") != Proto.OFFER) return

            val pin = String.format("%06d", Random().nextInt(1_000_000))
            val info = IncomingInfo(
                senderName = offer.optString("name", "?"),
                fileCount = offer.optInt("fileCount", 0),
                totalSize = offer.optLong("totalSize", 0),
                pin = pin
            )
            announced = true
            onIncoming(info)
            onLog("Solicitud de «${info.senderName}» (${info.fileCount} archivo/s). PIN: $pin")

            val expires = System.currentTimeMillis() + 60_000
            f.writeLine(msg(Proto.PIN_REQUIRED))

            val pinLine = f.readLine()
            val pinMsg = if (pinLine != null) JSONObject(pinLine) else null
            if (pinMsg == null || pinMsg.optString("type") != Proto.PIN) { f.writeLine(reject("denied")); return }
            if (System.currentTimeMillis() > expires) { f.writeLine(reject("expired")); onLog("PIN expirado."); return }
            if (pinMsg.optString("pin") != pin) { f.writeLine(reject("bad_pin")); onLog("PIN incorrecto."); return }

            f.writeLine(msg(Proto.ACCEPTED))

            val total = if (info.totalSize > 0) info.totalSize else 1L
            var done = 0L
            while (true) {
                val line = f.readLine() ?: break
                val m = JSONObject(line)
                val type = m.optString("type")
                if (type == Proto.DONE) { ok = true; break }
                if (type != Proto.FILE) continue

                val name = sanitize(m.optString("name", "archivo"))
                val size = m.optLong("size", 0)
                val base = done
                val (out, savedName) = openReceiveOutput(name)
                out.use { os ->
                    f.copyExact(os, size) { written ->
                        onProgress(minOf(1f, (base + written) / total.toFloat()))
                    }
                }
                done += size
                onLog("Recibido: $savedName (${formatSize(size)})")
            }
        } catch (e: Exception) {
            onLog("Error en recepción: ${e.message}")
        } finally {
            onProgress(0f)
            if (announced) onIncoming(null)
            try { client.close() } catch (_: Exception) {}
        }
    }

    // ---- Envío ----

    /** Elemento a enviar: nombre, tamaño y un abridor de flujo perezoso. */
    private class OutItem(val name: String, val size: Long, val open: () -> InputStream)

    /** Envía archivos elegidos con el selector del sistema (content URIs). */
    fun sendUris(peer: Peer, uris: List<Uri>, pinProvider: () -> String?): Boolean {
        val items = uris.mapNotNull { u ->
            queryMeta(u)?.takeIf { it.size > 0 }?.let { meta ->
                OutItem(meta.name, meta.size) { context.contentResolver.openInputStream(u)!! }
            }
        }
        return sendItems(peer, items, pinProvider)
    }

    /** Envía archivos elegidos con el explorador interno (java.io.File). */
    fun sendFiles(peer: Peer, files: List<File>, pinProvider: () -> String?): Boolean {
        val items = files.filter { it.isFile && it.length() > 0 }
            .map { file -> OutItem(file.name, file.length()) { FileInputStream(file) } }
        return sendItems(peer, items, pinProvider)
    }

    private fun sendItems(peer: Peer, items: List<OutItem>, pinProvider: () -> String?): Boolean {
        if (items.isEmpty()) { onLog("No hay archivos válidos que enviar."); return false }
        val total = items.sumOf { it.size }

        try {
            Socket().use { sock ->
                sock.connect(InetSocketAddress(peer.host, peer.port), 8000)
                val f = Framed(sock.getInputStream(), sock.getOutputStream())

                f.writeLine(
                    JSONObject()
                        .put("v", Proto.VERSION).put("type", Proto.OFFER)
                        .put("name", selfName).put("fileCount", items.size)
                        .put("totalSize", total).toString()
                )

                val reqLine = f.readLine()
                val req = if (reqLine != null) JSONObject(reqLine) else null
                if (req == null || req.optString("type") != Proto.PIN_REQUIRED) {
                    onLog("El receptor no solicitó PIN."); return false
                }

                val pin = pinProvider() ?: run { onLog("Envío cancelado."); return false }
                f.writeLine(
                    JSONObject().put("v", Proto.VERSION).put("type", Proto.PIN)
                        .put("pin", pin.trim()).toString()
                )

                val respLine = f.readLine()
                val resp = if (respLine != null) JSONObject(respLine) else null
                if (resp == null || resp.optString("type") != Proto.ACCEPTED) {
                    onLog("Rechazado por el receptor: ${resp?.optString("reason") ?: "desconocido"}")
                    return false
                }

                var done = 0L
                for (item in items) {
                    f.writeLine(
                        JSONObject().put("v", Proto.VERSION).put("type", Proto.FILE)
                            .put("name", item.name).put("size", item.size).toString()
                    )
                    val base = done
                    item.open().use { ins ->
                        f.writeBytes(ins, item.size) { sent ->
                            onProgress(minOf(1f, (base + sent) / total.toFloat()))
                        }
                    }
                    done += item.size
                    onLog("Enviado: ${item.name} (${formatSize(item.size)})")
                }
                f.writeLine(msg(Proto.DONE))
                onLog("Transferencia completada.")
                return true
            }
        } catch (e: Exception) {
            onLog("Error en envío: ${e.message}")
            return false
        } finally {
            onProgress(0f)
        }
    }

    private data class FileMeta(val name: String, val size: Long)

    private fun queryMeta(uri: Uri): FileMeta? {
        var name = "archivo"
        var size = 0L
        try {
            context.contentResolver.query(uri, null, null, null, null)?.use { c ->
                val ni = c.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                val si = c.getColumnIndex(OpenableColumns.SIZE)
                if (c.moveToFirst()) {
                    if (ni >= 0) c.getString(ni)?.let { name = it }
                    if (si >= 0 && !c.isNull(si)) size = c.getLong(si)
                }
            }
            if (size <= 0L) {
                context.contentResolver.openFileDescriptor(uri, "r")?.use {
                    if (it.statSize > 0) size = it.statSize
                }
            }
        } catch (_: Exception) {
            return null
        }
        return FileMeta(name, size)
    }

    private fun msg(type: String) =
        JSONObject().put("v", Proto.VERSION).put("type", type).toString()

    private fun reject(reason: String) =
        JSONObject().put("v", Proto.VERSION).put("type", Proto.REJECTED).put("reason", reason).toString()

    private fun sanitize(raw: String): String {
        val name = raw.substringAfterLast('/').substringAfterLast('\\')
            .replace(Regex("[\\\\/:*?\"<>|]"), "_").trim()
        return name.ifBlank { "archivo" }
    }

    /**
     * Abre el flujo de salida para un archivo entrante en la carpeta de recepción:
     * usa la carpeta SAF elegida por el usuario si existe; si no, [downloadDir].
     * Devuelve el flujo y el nombre final guardado.
     */
    private fun openReceiveOutput(name: String): Pair<OutputStream, String> {
        val uri = receiveTreeUri
        if (uri != null) {
            try {
                val tree = DocumentFile.fromTreeUri(context, uri)
                if (tree != null && tree.canWrite()) {
                    val doc = tree.createFile(mimeOf(name), name)
                    if (doc != null) {
                        val os = context.contentResolver.openOutputStream(doc.uri)
                        if (os != null) return Pair(os, doc.name ?: name)
                    }
                }
            } catch (_: Exception) {
                // Si falla la carpeta SAF, se cae a la carpeta interna.
            }
        }
        val file = uniqueFile(name)
        return Pair(FileOutputStream(file), file.name)
    }

    private fun mimeOf(name: String): String {
        val ext = name.substringAfterLast('.', "").lowercase()
        return MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext)
            ?: "application/octet-stream"
    }

    private fun uniqueFile(name: String): File {
        var f = File(downloadDir, name)
        if (!f.exists()) return f
        val dot = name.lastIndexOf('.')
        val base = if (dot > 0) name.substring(0, dot) else name
        val ext = if (dot > 0) name.substring(dot) else ""
        var i = 1
        while (true) {
            f = File(downloadDir, "$base ($i)$ext")
            if (!f.exists()) return f
            i++
        }
    }

    fun stop() {
        running = false
        try { serverSocket?.close() } catch (_: Exception) {}
    }
}
